bookchain-miner/src/mainpeers.cpp

```diff
@@ -0,0 +1,182 @@
+#include "mainpeers.hpp"
+#include "chain.hpp"
+#include "utils.hpp"
+#include "version.hpp"
+
+#include <curl/curl.h>
+#include <iostream>
+#include <nlohmann/json.hpp>
+#include <string>
+#include <thread>
+
+namespace bookchain {
+
+size_t stringWriteCallback(char* ptr, size_t size, size_t nmemb, std::string* userdata) {
+    size_t realSize = size * nmemb;
+    userdata->append(ptr, realSize);
+
+    return realSize;
+}
+
+std::string getRequest(const std::string& url) {
+    CURL* curl = nullptr;
+    CURLcode response;
+    std::string returnString;
+
+    curl = curl_easy_init();
+    if (curl != nullptr) {
+        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
+
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, stringWriteCallback);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &returnString);
+
+        response = curl_easy_perform(curl);
+        if (response != CURLE_OK) {
+            returnString = "";
+        }
+
+        curl_easy_cleanup(curl);
+    }
+
+    return (returnString.empty()) ? "{}" : returnString;
+}
+
+std::string postRequest(const std::string& url, const std::string& payload) {
+    CURL* curl = nullptr;
+    CURLcode response;
+    std::string returnString;
+
+    curl = curl_easy_init();
+    ;
+    if (curl != nullptr) {
+        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
+
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, payload.c_str());
+
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, stringWriteCallback);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &returnString);
+
+        response = curl_easy_perform(curl);
+        if (response != CURLE_OK) {
+            returnString = "";
+        }
+
+        curl_easy_cleanup(curl);
+    }
+
+    return (returnString.empty()) ? "{}" : returnString;
+}
+
+void discoverInitialPeers(PeersList& peersList) {
+    const std::array<std::string, 1> peerHints = {"127.0.0.1:8000"};
+
+    auto discoverPeer = [&peersList](const std::string& peerHint) {
```

Comment:
Can you explain the square bracket notation in this context for a c++ noob :)

Reply:
So this is a lambda as you've probably guessed. C++ doesn't capture variables from the outside scope into the lambda automatically for whatever technical reason. The square brackets are to explicitly capture variables and put them into the scope of the lambda.